extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -416,61 +421,87 @@ public void start()
             ioConfig.getStartDelay(),
             spec.toString()
         );
+
+        initializationExec.shutdownNow();
       }
       catch (Exception e) {
         if (consumer != null) {
           consumer.close();
         }
         log.makeAlert(e, "Exception starting KafkaSupervisor[%s]", dataSource)
            .emit();
-        throw Throwables.propagate(e);
       }
     }
   }
 
+  @Override
+  public void start()
+  {
+    synchronized (stateChangeLock) {
+      Preconditions.checkState(!lifecycleStarted, "already started");
+      Preconditions.checkState(!exec.isShutdown(), "already stopped");
+
+      // Try normal initialization first, if that fails then schedule periodic initialization retries
+      tryInit();
+      if (!started) {
+        initializationExec.scheduleAtFixedRate(
```

Comment:
How about using `RetryUtils.retry()`? It has a backoff mechanism which I think makes sense to apply here.

Reply:
Since there are checks for submitting an invalid supervisor spec (e.g., if nothing in "bootstrap.servers" is resolvable), the error condition this is addressing seems to be situations where some external state changes and the supervisor spec no longer works. For the unresolvable bootstrap.servers issue as an example, maybe its a transient DNS error, so I feel like it's better to retry indefinitely without increasing backoff times (with retries tied to the configured supervisor run period instead).

RetryUtils.retry() is also a blocking call, which I want to avoid here in the lifecycle startup.

Reply:
Hmm, I think backoff is more appropriate for transient errors. It starts with  retrying in a very short interval at first, but the retry interval grows as the failure continues. This makes sense because it can recover quickly with short retry periods if the error is transient. If the error is permanent, the retry period would grow in the end and it won't use excessive resources. Regarding too large backoff time, `RetryUtils.retry()` has the `MAX_SLEEP_MILLIS` which I think it makes sense to make it configurable. Regarding blocking call, please check my below comment. What do you think?

Reply:
Also, does it make sense to add a sort of max limit for retrying?

Reply:
Hm, I suppose the user can resubmit the supervisor in the situation where the backoff gets extremely large, I'll change to have limited retries with backoff